File: Beecrowd/4 - Data Structures and Libraries/1022 - ~/1022 - TDA Rational/1022_TDARational.py
import math


def gcd(a, b):
    return math.gcd(a, b)


def simplify(numerator, denominator):
    if denominator < 0:
        numerator = -numerator
        denominator = -denominator
    divisor = gcd(abs(numerator), abs(denominator))
    return numerator // divisor, denominator // divisor


def process_operation(n1, d1, n2, d2, op):
    if op == "+":
        result_numerator = n1 * d2 + n2 * d1
        result_denominator = d1 * d2
    elif op == "-":
        result_numerator = n1 * d2 - n2 * d1
        result_denominator = d1 * d2
    elif op == "*":
        result_numerator = n1 * n2
        result_denominator = d1 * d2
    elif op == "/":
        result_numerator = n1 * d2
        result_denominator = n2 * d1
    else:
        print("Invalid operator")
        return

    print(f"{result_numerator}/{result_denominator} ", end="")

    result_numerator, result_denominator = simplify(
        result_numerator, result_denominator
    )
    print(f"= {result_numerator}/{result_denominator}")


def main():
    import sys

    input = sys.stdin.read
    data = input().splitlines()

    T = int(data[0])
    for i in range(1, T + 1):
        line = data[i]
        parts = line.split()
        n1 = int(parts[0])
        d1 = int(parts[2])
        op = parts[3]
        n2 = int(parts[4])
        d2 = int(parts[6])

        process_operation(n1, d1, n2, d2, op)


if __name__ == "__main__":
    main()

File: Beecrowd/4 - Data Structures and Libraries/1022 - ~/1022 - TDA Rational/1022_TDARational.cpp
#include <bits/stdc++.h>

using namespace std;

int gcd(int a, int b) { return std::gcd(a, b); }

void simplify(int &numerator, int &denominator) {
    if (denominator < 0) {
        numerator = -numerator;
        denominator = -denominator;
    }
    
    int divisor = gcd(abs(numerator), abs(denominator));
    numerator /= divisor;
    denominator /= divisor;
}

void processOperation(int n1, int d1, int n2, int d2, char op) {
    int result_numerator, result_denominator;

    switch (op) {
    case '+':
        result_numerator = n1 * d2 + n2 * d1;
        result_denominator = d1 * d2;
        break;
    case '-':
        result_numerator = n1 * d2 - n2 * d1;
        result_denominator = d1 * d2;
        break;
    case '*':
        result_numerator = n1 * n2;
        result_denominator = d1 * d2;
        break;
    case '/':
        result_numerator = n1 * d2;
        result_denominator = n2 * d1;
        break;
    default:
        cerr << "Invalid operator" << endl;
        return;
    }

    cout << result_numerator << '/' << result_denominator << ' ';

    simplify(result_numerator, result_denominator);

    cout << "= " << result_numerator << '/' << result_denominator << endl;
}

int main() {
    int T;
    cin >> T;
    cin.ignore();

    for (int i = 0; i < T; i++) {
        int n1, d1, n2, d2;
        char op, slash1, slash2;

        cin >> n1 >> slash1 >> d1 >> op >> n2 >> slash2 >> d2;

        processOperation(n1, d1, n2, d2, op);
    }

    return 0;
}
